src/metabase/query_processor/pivot/postprocess.clj

```diff
@@ -491,13 +507,20 @@
             ;; filtering out any rows that begin with "Totals ..."
             (mapv
              (fn [row]
-               (let [[row-part vals-part] (split-at (count pivot-rows) row)
-                     first-entry (first row-part)]
-                 (if (or
-                      (not (seq row-part))
-                      (and (string? first-entry) (str/starts-with? first-entry "Totals")))
-                   row
-                   (into (mapv fmt row-formatters row-part) vals-part)))))))
+               (let [pivot-rows-cnt (count pivot-rows)]
+                 (if (pos? pivot-rows-cnt)
+                   (let [first-entry (first row)]
+                     (if (and (string? first-entry) (str/starts-with? first-entry "Totals"))
+                       row
+                       (let [it (.iterator ^Iterable row)]
+                         (loop [res (transient []), i 0]
+                           (if (.hasNext it)
+                             (recur (conj! res (if (< i pivot-rows-cnt)
+                                                 (fmt (nth row-formatters i) (.next it))
+                                                 (.next it)))
+                                    (unchecked-inc i))
+                             (persistent! res))))))
+                   row))))))
```

Comment:
Could you break this whole formatting step out as a separate function? This is getting pretty hairy

Reply:
A brief comment explaining the optimizations might also be helpful here